### PR TITLE
actually fixes blob spore pathfinding

### LINKED ISSUE #6098: Blob spores remain docile and passive
## Round ID:
35152

## Reproduction:
On the mentioned round ID a zombifying blob spawned. Its multiple spores were passive and never moved on their own to attack the crew.

### PR DESCRIPTION
**Pull request by MNarath1**
## About The Pull Request
In https://github.com/BeeStation/BeeStation-Hornet/pull/5967 i tried to fix blob pathfinding by simply replacing the normal pathfinding with jps pathfinding turned out trough there was more to  this than just that and some test cases that i did not think of at that time revealed problems that i am gonna adress with this PR.
closes https://github.com/BeeStation/BeeStation-Hornet/issues/6098
Edit:
The fallback for pathfinding works by making the assumtion that we have something blocking our way so we step from the target towards the origin until we find a turf that either is a ~~turf/closed~~ a turf with density true and no matching passflag or has an object on it that qualfies as a blocker when we find it we step 1 further and if that one is free so no blockers we try to pathfind towards that if we still fail the same procedure goes again until we stepped completly back towards the origin if we get there then pathfinding truely has failed.
~~Important Note trough not all possible blockers are adressed so there still might be scenarios that would have it fail (cause i can't just assume all machines as blockers cause pipes are machines too etc.)~~
This is a non issue after going away from checking their typpath and instead check if they are dense and if the passflags don't match that way we can adress any possible blocker.

## Why It's Good For The Game
FIX

## Changelog
:cl:
fix: fixed blob spore pathing
tweak: blob spores will keep hunting target mobs until those are out of range
/:cl:

### CODE REVIEW

**Comment by MNarath1** [comment]
also yes the fallback is not perfect but i feel looking for the blocker is a good enough compromise to find out if we can try to pathfind again and should reduce the actually needed calls for get_path_to in this proc a LOT and thats good considering its the most expensive proc in here.

**Comment by Penwin0** [comment]
Hero

File: code/modules/antagonists/blob/blob_mobs.dm [hunk 1]
```diff
@@ -208,16 +209,45 @@
...
 
+	if(!movement_steps) //pathfinding fallback in case we cannot find a valid path at the first attempt
+		var/ln = get_dist(src, target)
+		var/target_new = target
+		var/found_blocker
+		while(!movement_steps && (ln > 0)) //will stop if we can find a valid path or if ln gets reduced to 0 or less
+			for(var/i in 1 to ln) //calling get_path_to every time is quite taxing lets see if we can find whatever blocks us
+				target_new = get_step(target_new,  get_dir(target_new, src)) //step towards the origin until we find the blocker then 1 further
+				ln--
+				if(istype(target_new,/turf/closed) || (locate(/obj/structure/window) in target_new) || (locate(/obj/machinery/door) in target_new) || (locate(/obj/structure/table) in target_new) || (locate(/obj/mecha) in target_new)) //we check for possible things that could block us
```
Comment (MNarath1) [inline code/modules/antagonists/blob/blob_mobs.dm]: If anyone has a better idea how to do this part i am very much open to suggestions

Comment (MNarath1) [reply code/modules/antagonists/blob/blob_mobs.dm]: i mean line 237

Comment (Archanial) [reply code/modules/antagonists/blob/blob_mobs.dm]: for var/stuff in turf.contents if istype in list

Comment (MNarath1) [reply code/modules/antagonists/blob/blob_mobs.dm]: > for var/stuff in turf.contents if istype in list

Already tried that by benchmarking it for some reason it seems to be faster to do this with locate but that makes no sense at all honestly

**Comment by MNarath1** [comment]
i could technically do this with a big istype too but that one would really be slow

**Comment by MNarath1** [comment]
now using only istype means the blocker could be missed but in the end it should still be fine since this is just a fallback option and the assoc list was appearantly more expensive in memory and also kinda processing so i decided to go back to the istype solution

**Comment by MNarath1** [comment]
Now there are some uhhh concerns with the speed of the spores honestly someone else needs to take a look at this same for how far away you should be able to get before the spore gets its pathfinding canceled so it stops hunting you momentarly.

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**